presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java

```diff
@@ -300,8 +302,13 @@ private Expression coerceIfNecessary(Expression original, Expression rewritten)
         }, expression, null);
     }
 
-    Optional<Symbol> getSymbol(RelationPlan plan, Expression expression)
+    private Optional<Symbol> getSymbol(RelationPlan plan, Expression expression)
     {
+        if (!analysis.isColumnReference(expression)) {
```

Comment:
when this is needed?

Reply:
This is needed e.g. when `r` is a column reference of type `row(a: integer, ...)` and we try to get symbol for `r.a`.

Reply:
When this would happen and why it wasn't a problem before?
Scope shouldn't resolve non column reference fields with `.tryResolveField(expression)`. What test does fail?

Reply:
It would be redoing analysis in the query planning. Why would I want that?

Reply:
Well, using scope again is redoing analysis, but we have to live with this for now (this method will be gone eventually). However, as I mentioned, scope resolves only column references. That is why this change is questionable. In fact, the intention was to prevent using incorrect (non-lambda) scope for lambda arg reference. Because of such, it would be more natural to check if expression is not a lambda reference.

Reply:
I agree with you. Except that it's not possible to just "check if expression is not a lambda reference" without redoing analysis, so I'd argue that the presented solution is more natural. Are we bike shedding?

Reply:
> Except that it's not possible to just "check if expression is not a lambda reference"

What about adding method to:
```
boolean Analysis::isLambdaArgumentReference(Expression expression) {
  returns expression instanceof Identifier && lambdaArgumentReferences.contains(NodeRef.of((Identifier) identifier));
}
```
which would be similar to `isColumnReference(Expression expression)`
?

Reply:
Indeed, this would work. But also it would also make logic more complex. Instead of "is column reference so resolvable with scope" it would be "is not lambda reference, so probably column reference so resolvable with scope".

Reply:
> Instead of "is column reference so resolvable with scope"

This is actually: "is expression that is column reference resolvable with scope (than only resolves column references)", but then I think: "hey, if it's not a column reference then it shouldn't be resolved by scope anyway, so why this additional check"

The entire planner and analyzer is very complex piece of code. The less WTF code the better (you experienced that too). The code here deserves comment at least.

Reply:
> The code here deserves comment at least.

Done.